lib/pos_blockchain.go

```diff
@@ -0,0 +1,1134 @@
+	// First make sure we have a non-nil header
+	if block.Header == nil {
+		return RuleErrorNilBlockHeader
+	}
+
+	// Make sure we have a prevBlockHash
+	if block.Header.PrevBlockHash == nil {
+		return RuleErrorNilPrevBlockHash
+	}
+
+	// Timestamp validation
+	// First make sure we have a non-nil header
+	if block.Header == nil {
+		return RuleErrorNilBlockHeader
+	}
+
+	// Make sure we have a prevBlockHash
+	if block.Header.PrevBlockHash == nil {
+		return RuleErrorNilPrevBlockHash
+	}
+
+	// Timestamp validation
+	// TODO: Add support for putting the drift into global params.
+	if block.Header.TstampNanoSecs > uint64(time.Now().UnixNano())+bc.params.DefaultBlockTimestampDriftNanoSecs {
+		return RuleErrorPoSBlockTstampNanoSecsInFuture
+	}
+
+	// Header validation
+	if block.Header.Version != HeaderVersion2 {
+		return RuleErrorInvalidPoSBlockHeaderVersion
+	}
+
+	// Malformed block checks
+	// All blocks must have at least one txn
+	if len(block.Txns) == 0 {
+		return RuleErrorBlockWithNoTxns
+	}
+	// Must have non-nil TxnConnectStatusByIndex
+	if block.TxnConnectStatusByIndex == nil {
+		return RuleErrorNilTxnConnectStatusByIndex
+	}
+
+	// Must have TxnConnectStatusByIndexHash
+	if block.Header.TxnConnectStatusByIndexHash == nil {
+		return RuleErrorNilTxnConnectStatusByIndexHash
+	}
+
+	// Make sure the TxnConnectStatusByIndex matches the TxnConnectStatusByIndexHash
+	if !(HashBitset(block.TxnConnectStatusByIndex).IsEqual(block.Header.TxnConnectStatusByIndexHash)) {
+		return RuleErrorTxnConnectStatusByIndexHashMismatch
+	}
+
+	// Require header to have either vote or timeout QC
+	isTimeoutQCEmpty := block.Header.ValidatorsTimeoutAggregateQC.isEmpty()
+	isVoteQCEmpty := block.Header.ValidatorsVoteQC.isEmpty()
+	if isTimeoutQCEmpty && isVoteQCEmpty {
+		return RuleErrorNoTimeoutOrVoteQC
+	}
+
+	if !isTimeoutQCEmpty && !isVoteQCEmpty {
+		return RuleErrorBothTimeoutAndVoteQC
+	}
+
+	if len(block.Txns) == 0 {
+		return RuleErrorBlockWithNoTxns
+	}
+
+	if block.Txns[0].TxnMeta.GetTxnType() != TxnTypeBlockReward {
+		return RuleErrorBlockDoesNotStartWithRewardTxn
+	}
+
+	if block.Header.ProposerVotingPublicKey.IsEmpty() {
+		return RuleErrorInvalidProposerVotingPublicKey
+	}
+
+	if block.Header.ProposerPublicKey == nil || block.Header.ProposerPublicKey.IsZeroPublicKey() {
+		return RuleErrorInvalidProposerPublicKey
+	}
+
+	if block.Header.ProposerRandomSeedHash.isEmpty() {
+		return RuleErrorInvalidRandomSeedHash
+	}
+
+	merkleRoot := block.Header.TransactionMerkleRoot
+
+	// We always need to check the merkle root.
+	if merkleRoot == nil {
+		return RuleErrorNilMerkleRoot
+	}
+	computedMerkleRoot, _, err := ComputeMerkleRoot(block.Txns)
+	if err != nil {
+		return errors.Wrapf(err, "isProperlyFormedBlockPoS: Problem computing merkle root")
+	}
+	if !merkleRoot.IsEqual(computedMerkleRoot) {
+		return RuleErrorInvalidMerkleRoot
+	}
+
+	// TODO: What other checks do we need to do here?
+	return nil
```

Comment:
Some things from our previous discussion:
* Check that ProposerPublicKey corresponds to a valid validator based on our next-hour snapshot.
* Check that the QC is signed by 2/3rds majority of stake.
* The RandomSeedHash is determined by signing the previous RandomSeedHash with the validator's key. So that can/should be checked here and in other parts of the validation.
* You can also check that the ProposerVotePartialSignature is valid.

Reply:
- ProposerPublicKey check for non-orphans is handled in `hasValidBlockProposerPoS`. This validation for orphans is a work in progress as there's some slightly different logic we need for orphans.
- QC is signed by 2/3rds majority is handled in `isValidPoSQuorumCertificate`. This validation for orphans is a work in progress as there's some slightly different logic we need for orphans.
- Didn't know that about the random seed hash, will add.
- yep will add a proposer vote partial sig validation